### PR TITLE
BREAKING CHANGE: bump the minimum React version support to 16.14.0

### PR DESCRIPTION
**Pull request by zchenwei**
#### Description of changes

1. As we discussed earlier, there is no reason for us to support React versions before 16.14.0. This is lowest React version Studio supports and has the new JSX transform backported by React team.

2. Update our dev build to align with this minimum React version. This can prevent our release from potentially breaking customers apps that are still running on React before 17 if anything is missing from React 16 but get included in our release. e.g, #2160 

#### Issue #, if available

1. Many of our React related dependencies require a higher React version(>16.9), like testing library. It would be a pain for our library to continue with 16.8

2. Plus, React 16.8 does not support the new JSX transform which generates clean code and ends up reducing the bundle size. 

#### Description of how you validated changes

Testing against a sample app running on React 16.14.0, React 17 and React 18

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react/tsconfig.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "baseUrl": ".",
     "declaration": true,
     "esModuleInterop": true,
-    "jsx": "react",
+    "jsx": "react-jsx",
```
Comment (zchenwei) [inline packages/react/tsconfig.json:7]: This will turn on the new JSX transform for Typescript compiler.

File: packages/react/package.json [hunk 2]
```diff
@@ -109,8 +109,8 @@
     "eslint-plugin-react-hooks": "^4.4.0",
     "jest": "^27.0.4",
     "jest-matchmedia-mock": "^1.1.0",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2",
+    "react": "16.14.0",
```
Comment (zchenwei) [inline packages/react/package.json:112]: pin the dev build version to the minimum React version we support in ui react

**Comment by reesscot** [comment]
Should we go ahead and make this change on the branches with ui-react-core as well?

**Comment by calebpollman** [comment]
> Should we go ahead and make this change on the branches with ui-react-core as well?

That's a valid callout, needs to be done on the `in-app-messaging/main` branch. 

@zchenwei Should we also update the pinned version of `@types/react` in the root _package.json_ to point at a `react@16.14` compatible version as well?

**Comment by zchenwei** [comment]
> > Should we go ahead and make this change on the branches with ui-react-core as well?
> 
> That's a valid callout, needs to be done on the `in-app-messaging/main` branch.
> 
> @zchenwei Should we also update the pinned version of `@types/react` in the root _package.json_ to point at a `react@16.14` compatible version as well?

That's a good callout. I am actually going back and forth with it. One the one hand, the tricky thing is there is no one to one match relationship between `types/react` and `react. `types/react` is normally a bit ahead of `react`, so literally we have no idea which version of `types/react` we should target to pin for `react@16.14.0`. Since `react@16.14.0` is the last release of react 16, I think no new types will be release into `types/react@~16.14.0` other than bug fixes or improvement, which still benefit us. Thoughts?

**Review by joebuono — DISMISSED** [review]

**Comment by zchenwei** [comment]
> LGTM! Will this fix the duplicate hooks issues we see building projects in stackblitz?

What was that duplicate hooks issues?

**Review by reesscot — COMMENTED** [review]
Looks good, we just need to update the migration guide to callout the new minimum React version.

**Comment by zchenwei** [comment]
> Looks good, we just need to update the migration guide to callout the new minimum React version.

Will do once I get the next-release/main branch into mine

File: docs/src/pages/_app.page.tsx [hunk 3]
```diff
@@ -91,7 +91,7 @@ function MyApp({ Component, pageProps }) {
         <ThemeProvider theme={baseTheme} colorMode={colorMode}>
           {
             /* Remove after Global Nav launch START */
-            isGlobalNavEnabled ? (  /* Remove after Global Nav launch END */
+            isGlobalNavEnabled /* Remove after Global Nav launch END */ ? (
```
Comment (zchenwei) [inline docs/src/pages/_app.page.tsx:94]: Prettier auto change

File: package.json [hunk 4]
```diff
@@ -40,6 +40,7 @@
       "**/@radix-ui",
       "**/@radix-ui/**",
       "**/react",
+      "**/react-dom",
```
Comment (zchenwei) [inline package.json:43]: `react` and `react-dom` should be in pair.

File: package.json [hunk 5]
```diff
@@ -48,7 +49,8 @@
     ]
   },
   "resolutions": {
-    "@types/react": "17.0.47",
+    "@types/react": "16.14.0",
+    "@types/react-dom": "16.9.0",
```
Comment (zchenwei) [inline package.json:53]: There is no `16.14.x` for `@types/react-dom`. This is the latest minor in 16.

Comment (reesscot) [reply package.json:53]: Hmm, does this mean we need to do additional testing?

**Comment by zchenwei** [comment]
After discussion with @reesscot offline, we think it is risky to get this out because we do not fully understand how `@types/react@16` will affect users with apps running on a higher react version. It is not worth it taking the risk of making such changes that nobody complains about now

**Pull request closed without merging**